graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java

```diff
@@ -0,0 +1,386 @@
+ * Handles potentially multiple subscriptions through the same web socket.
+ */
+class GraphqlWSSubProtocol {
+    private static final Logger logger = LoggerFactory.getLogger(GraphqlWSSubProtocol.class);
+    private final GraphqlExecutor graphqlExecutor;
+    private final HashMap<String, ExecutionResultSubscriber> graphqlSubscriptions = new HashMap<>();
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    private static final TypeReference<Map<String, Object>> JSON_MAP =
+            new TypeReference<Map<String, Object>>() {};
+
+    private boolean connectionInitiated;
+
+    private final ServiceRequestContext ctx;
+    private final Map<String, Object> upgradeCtx;
+    private Map<String, Object> connectionCtx = ImmutableMap.of();
+
+    GraphqlWSSubProtocol(ServiceRequestContext ctx, GraphqlExecutor executor) {
+        upgradeCtx = GraphqlServiceContexts.graphqlContext(ctx);
+        graphqlExecutor = executor;
+        this.ctx = ctx;
+    }
+
+    /**
+     * Called when a binary frame is received. Binary frames are not supported by the graphql-ws protocol.
+     */
+    @Nullable
+    public void handleBinary(WebSocketWriter out) {
+        out.close(WebSocketCloseStatus.INVALID_MESSAGE_TYPE, "Binary frames are not supported");
+    }
+
+    /**
+     * Receives an event and returns a response if one should be sent.
+     */
+    @Nullable
+    public void handleText(String event, WebSocketWriter out) {
+        try {
+            final Map<String, Object> eventMap = parseJsonString(event, JSON_MAP);
+            final String type = toStringFromJson(eventMap.get("type"));
+            if (type == null) {
+                throw new WebSocketCloseException(4400);
+            }
+            final String id;
+
+            switch (type) {
+                case "connection_init":
+                    if (connectionInitiated) {
+                        // Already initiated, that's an error
+                        throw new WebSocketCloseException(4429);
+                    }
+                    final Object rawPayload = eventMap.get("payload");
+                    if (rawPayload != null) {
+                        connectionCtx = toMapFromJson(rawPayload);
+                    }
+                    connectionInitiated = true;
+                    writeConnectionAck(out);
+                    break;
+                case "ping":
+                    writePong(out);
+                    break;
+                case "pong":
+                    break;
+                case "subscribe":
+                    ensureInitiated();
+                    id = toStringFromJson(eventMap.get("id"));
+                    if (id == null) {
+                        throw new WebSocketCloseException(4400);
+                    }
+                    final Map<String, Object> payload = toMapFromJson(eventMap.get("payload"));
+                    try {
+                        if (graphqlSubscriptions.containsKey(id)) {
```

Comment:
Note: Because we don't put the ID into the map right away, there's a chance that multiple operations for the same ID can be received without an error.
However, it's not critical, let's handle it in another PR. 😉